substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1079 @@
+	}
+
+	// Sort and truncate candidate tickets, cleanup storage.
+	fn sort_and_truncate(candidates: &mut Vec<u128>, max_tickets: usize) -> u128 {
+		candidates.sort_unstable();
+		candidates[max_tickets..].iter().for_each(TicketsData::<T>::remove);
+		candidates.truncate(max_tickets);
+		candidates[max_tickets - 1]
+	}
+
+	/// Sort the tickets which belong to the epoch with the specified `epoch_tag`.
+	///
+	/// At most `max_segments` are taken from the `UnsortedSegments` structure.
+	///
+	/// The tickets of the removed segments are merged with the tickets on the `SortedCandidates`
+	/// which is then sorted an truncated to contain at most `MaxTickets` entries.
+	///
+	/// If all the entries in `UnsortedSegments` are consumed, then `SortedCandidates` is elected
+	/// as the next epoch tickets, else it is saved to be used by next calls of this function.
+	pub(crate) fn sort_segments(max_segments: u32, epoch_tag: u8, metadata: &mut TicketsMetadata) {
+		let unsorted_segments_count = metadata.unsorted_tickets_count.div_ceil(SEGMENT_MAX_SIZE);
+		let max_segments = max_segments.min(unsorted_segments_count);
+		let max_tickets = Self::epoch_length() as usize;
+
+		// Fetch the sorted candidates (if any).
+		let mut candidates = SortedCandidates::<T>::take().into_inner();
+
+		// There is an upper bound to check only if we already sorted the max number
+		// of allowed tickets.
+		let mut upper_bound = *candidates.get(max_tickets - 1).unwrap_or(&TicketId::MAX);
+
+		// Consume at most `max_segments` segments.
+		// During the process remove every stale ticket from `TicketsData` storage.
+		for segment_idx in (0..unsorted_segments_count).rev().take(max_segments as usize) {
+			let segment = UnsortedSegments::<T>::take(segment_idx);
+			metadata.unsorted_tickets_count -= segment.len() as u32;
+
+			// Push only ids with a value less than the current `upper_bound`.
+			for ticket_id in segment {
+				if ticket_id < upper_bound {
+					candidates.push(ticket_id);
+				} else {
+					TicketsData::<T>::remove(ticket_id);
+				}
+			}
+
+			// As we approach the tail of the segments buffer the `upper_bound` value is expected
+			// to decrease (fast). We thus expect the number of tickets pushed into the
+			// `candidates` vector to follow an exponential drop.
+			//
+			// Given this, sorting and truncating after processing each segment may be an overkill
+			// as we may find pushing few tickets more and more often. Is preferable to perform
+			// the sort and truncation operations only when we reach some bigger threshold
+			// (currently set as twice the capacity of `SortCandidate`).
+			//
+			// The more is the protocol's redundancy factor (i.e. the ratio between tickets allowed
+			// to be submitted and the epoch length) the more this check becomes relevant.
+			if candidates.len() > 2 * max_tickets {
+				upper_bound = Self::sort_and_truncate(&mut candidates, max_tickets);
+			}
+		}
+
+		if max_tickets != 0 {
```

Comment:
Not sure this if makes that much sense. As the epoch length should always be not zero.